### ISSUE TITLE
If the Content-Type header is not null, the Accept header is not checked for api version

### ISSUE THREAD

**Issue by kristofferkarlsson** [Microsoft/aspnet-api-versioning#338]
In MediaTypeApiVersionReader.Read, Content-Type is checked for version first -- if it's not null. The problem is that I want to provide the api version in the Accept header, which does not work if there is any content in the Content-Type header. The suggested behaviour is to call ReadAcceptHeader if ReadContentTypeHeader returns null.

**Comment by commonsensesoftware** [comment]
There are a couple of edge cases with versioning by media type. I think you've just hit one of them. The rules are basic and meant to be straight forward.

1. **Content-Type** - the `v={version}` parameter indicates the API version when sending content to the server with methods such as `POST`, `PUT`, `PATCH` and so on.
2. **Accept** - the `v={version}` parameter indicate the API version when no content is provided such as `GET`

While it's more that reasonable to send both `Content-Type` and `Accept` in a request that sends content to the server and gets a response, it's arguably not that common. Similarly, while the input API version and output API version can technically be different, it's not a particular common scenario and it can't be intrinsically rationalized on the server side (at least in the ASP.NET stack). The initial action mapping handles the incoming content. If you wanted to handle response content in an API version different from the request API version, you'd have to handle the `Accept` header yourself. This can be achieved using a custom **MediaTypeFormatter** in Web API or a custom **IOutputFormatter** in ASP.NET Core.

If the client is not sending any content, then it should **not** contain the `Content-Type` header. When the client does send content, `Content-Type` always takes precedence over `Accept` because only the incoming request can be bound to an action. Alternate output in this scenario is something you'd have to customize, if you need it.

The other edge case is the `DELETE` method. This method is typically not expected to have `Accept`, `Content-Type`, or any other suitable header for that matter. This is currently a known limitation and the recommendation is to send `Accept` even though there will be no content. This is the most compelling scenario to have version-neutral actions, which is not currently supported. This is something I certainly want to enable in the future.

I hope that helps clear things up.

**Comment by commonsensesoftware** [comment]
Did this clear things up for you or are you still facing an issue?

**Comment by peteraritchie** [comment]
At least to me, using Content-Type to specify the API version when Accept exists is not what I would expect.

`Content-Type` is the media type of the _request_, and `Accept` is the acceptable media types of the _response_.

If using MediaTypeApiVersionReader, you're explicitly versioning the media type, not the "API version".

If my API needs to support versions of requests AND versions of responses, MediaTypeApiVersionReader currently doesn't allow for that. :(

**Comment by commonsensesoftware** [comment]
Yes - that's true. This is because there is no way to rationalize that in the platform. This is a more of a limitation of how HTTP is mapped to ASP.NET Core. If `Content-Type` and `Accept` don't match, there's no action that can really be taken. A single action is mapped to request (e.g. input **and** output).

You _could_ potential/partial solve this issue with custom media type formatters, but then versioning would be of media type formatters rather than controllers and actions. That would be fine if the controller and action logic is the same and only the payload is different.

From a REST perspective, I full agree that the input and output media types, even with versions, do not have to match. I'm not sure how this could be implemented on the ASP.NET stack (any flavor). You can do this today purely with media types (say JSON in and XML out), but not with versioning, which also controls what implementation is executed.

If you have ideas or suggestions about how seperate input and output execution logic could be handled, I'd listen to it.
